woocommerce/woocommerce-ios
# Settings > Experimental Features: add back the Products switch for M2 features

Part 1 (Show M2 switch in settings, and enable M2 features based on the switch value) of #2042 

## Changes

- In `BetaFeaturesViewController`, uncommented the code that displays a Products feature switch in Settings > Experimental Features
- Renamed `ProductsVisibility*` to `ProductsFeatureSwitch*` in `BetaFeaturesViewController` and `AppSettingsStore` for the previous feature switch that used to control the Products tab visibility and now control the publicly available experimental Products M2 features
- Updated the Products feature switch details copy in `BetaFeaturesViewController`
- Created `ProductDetailsFactory` that creates a product details view controller async based on the Products feature switch from app settings and other feature flags
- Updated all the existing cases where we create a product details screen with `ProductDetailsFactory` in `ProductLoaderViewController` (tapping on a product outside of Products tab), `ProductsViewController` (Products tab), and `ProductSearchUICommand` (tapping on a product from search)
- In `ProductFormViewController` and its related classes, DI'ed the release 2/3 availability bool values instead of reading from `featureFlagService`

## Testing

Prerequisites: the site has a simple product and non-simple product (e.g. variable), and an order with a simple product

- Go to Settings > Experimental Features, and turn off the Products feature switch
- Go to the Orders tab
- Tap on an order with a simple product
- Tap on the simple product under the Products section --> there should be no M2 features (no "..." in the navigation bar for product settings, no short description, no image editing)
- Go to Settings > Experimental Features, and turn on the Products feature switch
- Go back to the order with a simple product and open the product again --> M2 features should be available now ("..." in the navigation bar for product settings, short description, image editing)
- Go to Products tab and search for a non-simple product
- Tap on the non-simple product --> the readonly product details screen should be shown

## Example screenshots

Please note that the copy update on the Products tab WIP banner will be in a separate PR!

Dark | Light
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-05-13 at 18 15 27](https://user-images.githubusercontent.com/1945542/81800762-ebd35880-9545-11ea-95e0-aa91fa8607cd.png) | ![Simulator Screen Shot - iPhone 11 - 2020-05-13 at 18 16 28](https://user-images.githubusercontent.com/1945542/81800766-eece4900-9545-11ea-9767-55aec29888b8.png)


Update release notes:
